File: Decode_String/DecodeString.py
def decode_string(s: str) -> str:
    stack = []
    current_num = 0
    current_str = ""

    for char in s:
        if char.isdigit():
            current_num = current_num * 10 + int(char)
        elif char == "[":
            stack.append((current_num, current_str))
            current_num = 0
            current_str = ""
        elif char == "]":
            num, prev_str = stack.pop()
            current_str = prev_str + num * current_str
        else:
            current_str += char

    return current_str

# Ví dụ sử dụng:
s1 = "3[a]2[bc]"
s2 = "3[a2[c]]"
s3 = "2[abc]3[cd]ef"

print(decode_string(s1))  # Output: "aaabcbc"
print(decode_string(s2))  # Output: "accaccacc"
print(decode_string(s3))  # Output: "abcabccdcdcdef"

File: Decode_String/DecodeString.cpp
#include <iostream>
#include <stack>
#include <string>

std::string decodeString(const std::string& s) {
    std::stack<int> numStack;
    std::stack<std::string> strStack;
    std::string currentStr = "";
    int currentNum = 0;

    for (char c : s) {
        if (std::isdigit(c)) {
            currentNum = currentNum * 10 + (c - '0');
        } else if (c == '[') {
            numStack.push(currentNum);
            strStack.push(currentStr);
            currentNum = 0;
            currentStr = "";
        } else if (c == ']') {
            int repeatTimes = numStack.top();
            numStack.pop();
            std::string prevStr = strStack.top();
            strStack.pop();
            for (int i = 0; i < repeatTimes; ++i) {
                prevStr += currentStr;
            }
            currentStr = prevStr;
        } else {
            currentStr += c;
        }
    }

    return currentStr;
}

int main() {
    std::string s1 = "3[a]2[bc]";
    std::string s2 = "3[a2[c]]";
    std::string s3 = "2[abc]3[cd]ef";

    std::cout << decodeString(s1) << std::endl;  // Output: "aaabcbc"
    std::cout << decodeString(s2) << std::endl;  // Output: "accaccacc"
    std::cout << decodeString(s3) << std::endl;  // Output: "abcabccdcdcdef"

    return 0;
}
